### PR TITLE
add a test for keyword transformation #3910

### LINKED ISSUE #3910: xtdb.next.jdbc should transform/keywordize nested return values
e.g. with the `:builder-fn` [configured](https://github.com/xtdb/xtdb/pull/3886/files#diff-23eaea0515eaa95e7e73b79d824b6cea2b1676e3491e587ad1ef2857a5e3da85R41-R47) the following is observed:

```
SELECT p.*, 
       NEST_MANY(
           SELECT bs.bdr$submission$state
           FROM bdr_submission FOR ALL SYSTEM_TIME bs
           WHERE bs.project$identifier = p._id
           
       ) AS submission_history
FROM project p limit 5
=>
{:submission-history
 [{"bdr$submission$state" "Submission complete"}
  {"bdr$submission$state" "Submission in progress"}],
 :project/comment nil,
 :project/start-date #inst "1984-10-31T13:00:00.000-00:00",
 :project/date-modified nil,
 :project/contact-point nil,
 :project/name "Quadra Forest Block",
 :xt/id #uuid "0004f309-728f-479d-a93d-3743ac5ea9dc",
 :project/description nil,
 :project/date-created #inst "2011-08-15T14:00:00.000-00:00",
 :project/vba-identifier 132,
 :project/license nil,
 :project/status "Published",
 :project/end-date #inst "1984-10-31T13:00:00.000-00:00"}
```

### PR DESCRIPTION
**Pull request by mpisanko**
resolves #3910

### CODE REVIEW

**Comment by mpisanko** [comment]
happy to drop the second commit - just thought it might be nice for users to have a shorthand.

**Comment by jarohen** [comment]
@mpisanko I'm fine with both here, but let's just call it `builder-fn` and document what it does separately - that name's quite a mouthful :sweat_smile:

**Comment by mpisanko** [comment]
a side effect is that nil columns are now returned - if that's ok will squash these commits, otherwise easy to take it out.

File: jdbc/src/main/clojure/xtdb/next/jdbc.clj [hunk 1]
```diff
@@ -47,6 +58,16 @@
 
 (def col-reader (->col-reader :kebab-case-keyword))
 
+
+(def ^{:doc "Builder function which converts result-set rows' map keys to kebab-cased keywords (iteratively)"}
+  builder-fn
+  (nj-rs/as-maps-adapter
+   (fn [rs opts]
+     (nj-rs/as-unqualified-modified-maps
```
Comment (jarohen) [inline jdbc/src/main/clojure/xtdb/next/jdbc.clj:65]: let's stick with the one from optional (as the test-util was) - we're looking to round-trip documents as they were inserted, and I think I'd want to err on the side of dropping null values than inserting them (although it's obviously a subjective decision).

File: jdbc/src/main/clojure/xtdb/next/jdbc.clj [hunk 2]
```diff
@@ -20,11 +20,22 @@
...
+            (if (keyword? (first (keys m)))
+              m
+              (update-keys m #(.denormalize key-fn %))))]
+    (w/prewalk (fn [v]
+                 (cond-> v
+                   (instance? Map v)
+                   process-map
+
+                   (instance? Collection v)
+                   (->>
```
Comment (jarohen) [inline jdbc/src/main/clojure/xtdb/next/jdbc.clj]: in theory, given it's a pre-walk, we should be able to just have `(instance? List v) vec` here, and have the pre-walk deal with iterating it?

(also worth us adding `(instance? Set v) set`, if the cause of this one was prewalk not recognising Java collections)

File: jdbc/src/main/clojure/xtdb/next/jdbc.clj [hunk 3]
```diff
@@ -20,11 +20,22 @@
                    (String. StandardCharsets/UTF_8)))))
 
 (defn- denormalize-keys [v, ^IKeyFn key-fn]
-  (w/prewalk (fn [v]
-               (cond-> v
-                 (instance? Map v)
-                 (update-keys #(.denormalize key-fn %))))
-             v))
+  (letfn [(process-map [m]
+            (if (keyword? (first (keys m)))
```
Comment (jarohen) [inline jdbc/src/main/clojure/xtdb/next/jdbc.clj:24]: what's the reasoning (/assumption) behind only checking one key of the map here?

**Review by jarohen — CHANGES_REQUESTED** [review]

File: jdbc/src/main/clojure/xtdb/next/jdbc.clj (hunk 3 above)
Comment (mpisanko) [reply jdbc/src/main/clojure/xtdb/next/jdbc.clj:24]: resolved by delegating to prewalk which in practice confirmed your theory :)

File: jdbc/src/main/clojure/xtdb/next/jdbc.clj [hunk 4]
```diff
@@ -20,11 +20,21 @@
...
+    (w/prewalk (fn [v]
+                 (cond-> v
+                   (instance? Map v)
+                   (-> (update-keys #(.denormalize key-fn %)))
+
+                   (instance? List v)
+                   vec
+
+                   (instance? Set v)
+                   vec))
```
Comment (jarohen) [inline jdbc/src/main/clojure/xtdb/next/jdbc.clj]: nit: `set`

File: jdbc/src/main/clojure/xtdb/next/jdbc.clj [hunk 5]
```diff
@@ -20,11 +20,21 @@
...
-                 (update-keys #(.denormalize key-fn %))))
-             v))
+  (letfn [(process-map [m]
+            (if (keyword? (first (keys m)))
+              m
+              (update-keys m #(.denormalize key-fn %))))]
+    (w/prewalk (fn [v]
+                 (cond-> v
+                   (instance? Map v)
+                   (-> (update-keys #(.denormalize key-fn %)))
```
Comment (jarohen) [inline jdbc/src/main/clojure/xtdb/next/jdbc.clj]: nit: don't need `->` here

**Review by jarohen — APPROVED** [review]
with the two remaining nits addressed, good to go - thanks :pray:

**Pull request merged**